libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Core.hs

```diff
@@ -103,3 +96,27 @@ withCtx ctx m = do
   case runSpecTransM ctx m of
     Right x -> pure x
     Left e -> throwError e
+
+-- | OpaqueErrorString behaves like unit in comparisons, but contains an
+-- error string that can be displayed.
+newtype OpaqueErrorString = OpaqueErrorString (NonEmpty Text)
+  deriving (Generic, Show, Semigroup)
+
+instance Eq OpaqueErrorString where
+  _ == _ = True
```

Comment:
What's going on with this terrible implementation of Eq instance that violates referential transparency?

Reply:
The idea at first was to have this as a temporary workaround to make all predicate failures equal, and then once we have some decent notion of predicate failures on the spec side, then this could easily be changed to that type instead. 

It doesn't seem like we're going to actually add predicate failures to the spec in the nearby future, so I guess this was a case of YAGNI. I plan to get rid of this in some future PR where I'm going to refactor the whole conformance testing framework. I think right now that's not too important, because the current approach works fine, even if it's a bit ugly.